File: app/modules/scoring/domain/calculations/rsi.py
"""RSI (Relative Strength Index) calculation functions.

Pure calculation functions for RSI using pandas-ta.
"""

from typing import Optional

import numpy as np
import pandas as pd
import pandas_ta as ta

from app.modules.scoring.domain.constants import RSI_LENGTH


def calculate_rsi(closes: np.ndarray, length: int = RSI_LENGTH) -> Optional[float]:
    """
    Calculate Relative Strength Index.

    Args:
        closes: Array of closing prices
        length: RSI period (default 14)

    Returns:
        Current RSI value (0-100) or None if insufficient data
    """
    if len(closes) < length + 1:
        return None

    series = pd.Series(closes)
    rsi = ta.rsi(series, length=length)

    if rsi is not None and len(rsi) > 0 and not pd.isna(rsi.iloc[-1]):
        return float(rsi.iloc[-1].item())

    return None

File: trader-go/pkg/formulas/rsi.go
package formulas

import (
	"github.com/markcheno/go-talib"
)

// CalculateRSI calculates the Relative Strength Index
// Faithful translation from Python: app/modules/scoring/domain/calculations/rsi.py
//
// RSI Formula:
//   RSI = 100 - (100 / (1 + RS))
//   where RS = Average Gain / Average Loss over N periods
//
// Args:
//   closes: Array of closing prices
//   length: RSI period (typically 14)
//
// Returns:
//   Current RSI value (0-100) or nil if insufficient data
func CalculateRSI(closes []float64, length int) *float64 {
	if len(closes) < length+1 {
		return nil
	}

	// Use go-talib for RSI calculation
	rsi := talib.Rsi(closes, length)

	// Return the last value
	if len(rsi) > 0 && !isNaN(rsi[len(rsi)-1]) {
		result := rsi[len(rsi)-1]
		return &result
	}

	return nil
}

// isNaN checks if a float64 is NaN
func isNaN(f float64) bool {
	return f != f
}
